src/lib/support/CHIPSparseArray.h

```diff
@@ -0,0 +1,193 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    Copyright (c) 2016-2017 Nest Labs, Inc.
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <stddef.h>
+
+#include <core/CHIPEncoding.h>
+#include <core/CHIPError.h>
+#include <support/Base64.h>
+#include <support/CHIPMem.h>
+#include <support/CodeUtils.h>
+#include <support/SafeInt.h>
+#include <transport/raw/MessageHeader.h>
+
+namespace chip {
+
+template <uint16_t Capacity>
+class SparseU64Array
+{
+
+public:
+    SparseU64Array() { CHIP_ZERO_AT(mArray); }
+    /**
+     * @brief
+     *   Serialize the sparse array into a base 64 encoded string.
+     *   Only the values till mNextAvailable index are encoded.
+     *   The empty indexes between 0, and mNextAvailable, are also
+     *   encoded.
+     *
+     * @param[in] buf Buffer where serialized string is written
+     * @param[in] buflen Length of buf
+     * @return pointer to buf, or nullptr in case of error
+     */
+    const char * Serialize(char * buf, uint16_t & buflen)
+    {
+        char * out = nullptr;
+        size_t len = sizeof(uint64_t) * mNextAvailable;
+        VerifyOrExit(buflen >= SerializedSize(), buflen = SerializedSize());
+        VerifyOrExit(buf != nullptr, buflen = SerializedSize());
+        VerifyOrExit(CanCastTo<uint16_t>(len), buflen = SerializedSize());
```

Comment:
This doesn't make sense to me: this would indicate to the caller that they need to call again with the new buffer length, but that won't help.

It would make more sense to me to just `static_assert(UINT16_MAX / sizeof(uint64_t) > Capacity)` or so, which would make it safe to assign `sizeof(uint64_t) * mNextAvailable` to `uint16_t`, right?

Reply:
Yes, I agree. I was not too happy with this check either. I'll update the code..